pkg/streamingpromql/operators/and_binary_operation.go

```diff
@@ -0,0 +1,292 @@
+		return nil, nil
+	}
+
+	rightMetadata, err := a.Right.SeriesMetadata(ctx)
+	if err != nil {
+		return nil, err
+	}
+
+	defer types.PutSeriesMetadataSlice(rightMetadata)
+
+	if len(rightMetadata) == 0 {
+		// We can't produce any series, we are done.
+		return nil, nil
+	}
+
+	groupMap := map[string]*andGroup{}
+	groupKeyFunc := vectorMatchingGroupKeyFunc(a.VectorMatching)
+
+	// Iterate through the left-hand series, and create groups for each based on the matching labels.
+	a.leftSeriesGroups = make([]*andGroup, 0, len(leftMetadata))
+
+	for _, s := range leftMetadata {
+		groupKey := groupKeyFunc(s.Labels)
+		group, exists := groupMap[string(groupKey)] // Important: don't extract the string(...) call here - passing it directly allows us to avoid allocating it.
+
+		if !exists {
+			group = &andGroup{lastRightSeriesIndex: -1}
+			groupMap[string(groupKey)] = group
+		}
+
+		group.leftSeriesCount++
+		a.leftSeriesGroups = append(a.leftSeriesGroups, group)
+	}
+
+	// Iterate through the right-hand series, and find groups for each based on the matching labels.
+	a.rightSeriesGroups = make([]*andGroup, 0, len(rightMetadata))
+	outputSeriesCount := 0
+
+	for idx, s := range rightMetadata {
+		groupKey := groupKeyFunc(s.Labels)
+		group, exists := groupMap[string(groupKey)] // Important: don't extract the string(...) call here - passing it directly allows us to avoid allocating it.
+
+		if exists {
+			if group.lastRightSeriesIndex == -1 {
+				// First time a right-hand series has matched this group.
+				// We'll return all left-hand series matching this group, so add the count to the running total of output series.
+				outputSeriesCount += group.leftSeriesCount
+			}
+
+			group.lastRightSeriesIndex = idx
+		}
+
+		// Even if there is no matching group, we want to store a nil value here so we know to throw the series away when we read it later.
+		a.rightSeriesGroups = append(a.rightSeriesGroups, group)
+	}
+
+	// Iterate through the left-hand series again, and build the list of output series based on those that matched at least one series on the right.
+	outputSeries := make([]types.SeriesMetadata, 0, outputSeriesCount)
+
+	for seriesIdx, group := range a.leftSeriesGroups {
+		if group.lastRightSeriesIndex == -1 {
+			// This series doesn't match any series from the right side.
+			// Discard the group.
+			a.leftSeriesGroups[seriesIdx] = nil
+		} else {
+			outputSeries = append(outputSeries, leftMetadata[seriesIdx])
+		}
+	}
+
+	return outputSeries, nil
+}
+
+func (a *AndBinaryOperation) NextSeries(ctx context.Context) (types.InstantVectorSeriesData, error) {
+	for {
+		if len(a.leftSeriesGroups) == 0 {
+			// No more series to return.
+			return types.InstantVectorSeriesData{}, types.EOS
+		}
+
+		thisSeriesGroup := a.leftSeriesGroups[0]
+		a.leftSeriesGroups = a.leftSeriesGroups[1:]
+
+		if thisSeriesGroup == nil {
+			// The next series from the left side has no matching series on the right side.
```

Comment:
This is a bit confusing, because `a.Left.NextSeries(ctx)` will be the series at the same index as `thisSeriesGroup`. Really it's "this series", not "the next series".

I suggest to move reading the next series to outside of this check (also then not needing it on line 157). The if condition can still put the series back and continue.

Reply:
I've addressed the first part in 26fb606.

I'd prefer not to do the second part (moving the reading of the next left series to one place), as that will mean we're holding more series in memory for longer while we read the right series. It may also mean that downstream operators on the left side are holding more intermediate state while the right side is evaluated, which would again increase peak memory consumption.

Reply:
Actually, instead of reading it to just discard it, what do you think about extending `InstantVectorOperator` to have a `SkipSeries` method that moves the selector forward so we don't have to get slices for all the points etc.

Reply:
I had the same idea, I've already added something like this to our list of performance improvement ideas 🙂 I think this makes sense, but this is a far bigger change outside the scope of this PR (and I suspect outside the scope of our initial focus of getting MQE up and running).